File: vibe/project_types/detector.py
"""Enhanced project type detection with comprehensive framework support."""

import json
from pathlib import Path
from typing import Any


class ProjectDetector:
    """Intelligent project type detection based on files, dependencies,
    and structure.
    """

    def __init__(self, project_path: str = "."):
        """Initialize the project detector with a given path."""
        self.project_path = Path(project_path)

    def detect_project_type(self) -> str:
        """Detect the primary project type."""
        detections = self.detect_all_project_types()

        # Return the most specific/confident detection
        # Priority order: specific frameworks > language > generic
        priority_order = [
            "vue",
            "react",
            "next",
            "nuxt",
            "svelte",
            "angular",
            "django",
            "fastapi",
            "flask",
            "rails",
            "typescript",
            "javascript",
            "python",
            "rust",
            "go",
            "web",
            "node",
            "generic",
        ]

        for project_type in priority_order:
            if project_type in detections:
                return project_type

        return "generic"

    def detect_all_project_types(self) -> list[str]:
        """Detect all applicable project types/frameworks."""
        types = []

        # Detect JavaScript/TypeScript frameworks and add their tech stack
        types.extend(self._detect_frontend_frameworks())

        # Detect Python and its frameworks
        types.extend(self._detect_python_stack())

        # Detect other language projects
        types.extend(self._detect_other_languages())

        # Detect general project characteristics
        types.extend(self._detect_general_types())

        # Remove duplicates while preserving order
        return list(dict.fromkeys(types))

    def _detect_frontend_frameworks(self) -> list[str]:
        """Detect frontend frameworks and their associated tech stack."""
        types = []

        if self._is_vue_project():
            types.extend(["vue", "javascript", "web", "node"])

        if self._is_react_project():
            types.extend(["react", "javascript", "web", "node"])

        if self._is_next_project():
            types.extend(["next", "react", "javascript", "web", "node"])

        if self._is_nuxt_project():
            types.extend(["nuxt", "vue", "javascript", "web", "node"])

        if self._is_svelte_project():
            types.extend(["svelte", "javascript", "web", "node"])

        if self._is_angular_project():
            types.extend(["angular", "typescript", "web", "node"])

        if self._is_typescript_project():
            types.extend(["typescript", "javascript", "node"])

        if self._is_javascript_project():
            types.extend(["javascript", "node"])

        return types

    def _detect_python_stack(self) -> list[str]:
        """Detect Python and its frameworks."""
        types = []

        if self._is_python_project():
            types.append("python")

            # Check for Python frameworks
            if self._is_django_project():
                types.append("django")
            elif self._is_fastapi_project():
                types.append("fastapi")
            elif self._is_flask_project():
                types.append("flask")

        return types

    def _detect_other_languages(self) -> list[str]:
        """Detect other programming languages."""
        types = []

        if self._is_rust_project():
            types.append("rust")

        if self._is_go_project():
            types.append("go")

        return types

    def _detect_general_types(self) -> list[str]:
        """Detect general project characteristics."""
        types = []

        if self._is_web_project():
            types.append("web")

        return types

    def _is_vue_project(self) -> bool:
        """Detect Vue.js projects."""
        # Check for Vue files
        if self._has_files_with_extension(".vue"):
            return True

        # Check package.json for Vue dependencies
        if self._has_npm_dependency(["vue", "@vue/cli", "vite", "nuxt"]):
            return True

        return False

    def _is_react_project(self) -> bool:
        """Detect React projects."""
        if self._has_npm_dependency(["react", "react-dom", "create-react-app"]):
            return True

        if self._has_files_with_extension([".jsx", ".tsx"]):
            return True

        return False

    def _is_next_project(self) -> bool:
        """Detect Next.js projects."""
        return self._has_npm_dependency(["next"]) or self._file_exists("next.config.js")

    def _is_nuxt_project(self) -> bool:
        """Detect Nuxt projects."""
        return self._has_npm_dependency(["nuxt", "@nuxt/cli"]) or self._file_exists(
            "nuxt.config.js"
        )

    def _is_svelte_project(self) -> bool:
        """Detect Svelte projects."""
        return self._has_npm_dependency(
            ["svelte", "@sveltejs/kit"]
        ) or self._has_files_with_extension(".svelte")

    def _is_angular_project(self) -> bool:
        """Detect Angular projects."""
        return self._has_npm_dependency(
            ["@angular/core", "@angular/cli"]
        ) or self._file_exists("angular.json")

    def _is_typescript_project(self) -> bool:
        """Detect TypeScript projects."""
        if self._file_exists("tsconfig.json"):
            return True

        if self._has_npm_dependency(["typescript"]):
            return True

        if self._has_files_with_extension([".ts", ".tsx"]):
            return True

        return False

    def _is_javascript_project(self) -> bool:
        """Detect JavaScript/Node.js projects."""
        if self._file_exists("package.json"):
            return True

        if self._has_files_with_extension([".js", ".mjs", ".cjs"]):
            return True

        return False

    def _is_python_project(self) -> bool:
        """Detect Python projects."""
        python_files = [
            "pyproject.toml",
            "setup.py",
            "requirements.txt",
            "Pipfile",
            "poetry.lock",
            "setup.cfg",
        ]

        for file in python_files:
            if self._file_exists(file):
                return True

        if self._has_files_with_extension(".py"):
            return True

        return False

    def _is_django_project(self) -> bool:
        """Detect Django projects."""
        if self._has_python_dependency(["django", "Django"]):
            return True

        if self._file_exists("manage.py"):
            return True

        return False

    def _is_fastapi_project(self) -> bool:
        """Detect FastAPI projects."""
        return self._has_python_dependency(["fastapi", "uvicorn"])

    def _is_flask_project(self) -> bool:
        """Detect Flask projects."""
        return self._has_python_dependency(["flask", "Flask"])

    def _is_rust_project(self) -> bool:
        """Detect Rust projects."""
        return self._file_exists("Cargo.toml")

    def _is_go_project(self) -> bool:
        """Detect Go projects."""
        return self._file_exists("go.mod") or self._has_files_with_extension(".go")

    def _is_web_project(self) -> bool:
        """Detect general web projects."""
        web_files = ["index.html", "style.css", "styles.css", "main.css"]
        for file in web_files:
            if self._file_exists(file):
                return True

        if self._has_files_with_extension([".html", ".css", ".scss", ".sass"]):
            return True

        return False

    def _file_exists(self, filename: str) -> bool:
        """Check if a file exists in the project root."""
        return (self.project_path / filename).exists()

    def _has_files_with_extension(self, extensions: str | list[str]) -> bool:
        """Check if files with given extensions exist."""
        if isinstance(extensions, str):
            extensions = [extensions]

        for ext in extensions:
            if not ext.startswith("."):
                ext = "." + ext

            # Search for files with this extension
            for file_path in self.project_path.rglob(f"*{ext}"):
                # Skip node_modules, __pycache__, etc.
                if not any(
                    part.startswith(".")
                    or part in ["node_modules", "__pycache__", "venv", "env"]
                    for part in file_path.parts
                ):
                    return True

        return False

    def _has_npm_dependency(self, dependencies: list[str]) -> bool:
        """Check if any of the dependencies exist in package.json."""
        package_json_path = self.project_path / "package.json"
        if not package_json_path.exists():
            return False

        try:
            with open(package_json_path) as f:
                package_data = json.load(f)

            all_deps = {}
            all_deps.update(package_data.get("dependencies", {}))
            all_deps.update(package_data.get("devDependencies", {}))
            all_deps.update(package_data.get("peerDependencies", {}))

            for dep in dependencies:
                if dep in all_deps:
                    return True

        except (OSError, json.JSONDecodeError):
            pass

        return False

    def _has_python_dependency(self, dependencies: list[str]) -> bool:
        """Check if any of the dependencies exist in Python project files."""
        return self._check_pyproject_dependencies(
            dependencies
        ) or self._check_requirements_dependencies(dependencies)

    def _check_pyproject_dependencies(self, dependencies: list[str]) -> bool:
        """Check dependencies in pyproject.toml file."""
        pyproject_path = self.project_path / "pyproject.toml"
        if not pyproject_path.exists():
            return False

        # Try structured parsing first
        if self._check_pyproject_structured(pyproject_path, dependencies):
            return True

        # Fallback to text parsing
        return self._check_pyproject_text(pyproject_path, dependencies)

    def _check_pyproject_structured(
        self, pyproject_path: Path, dependencies: list[str]
    ) -> bool:
        """Check pyproject.toml using structured TOML parsing."""
        try:
            import tomllib

            with open(pyproject_path, "rb") as f:
                pyproject_data = tomllib.load(f)

            # Check dependencies in project section
            project_deps = pyproject_data.get("project", {}).get("dependencies", [])
            for dep_line in project_deps:
                dep_name = self._extract_dependency_name(dep_line)
                if dep_name.lower() in [d.lower() for d in dependencies]:
                    return True

        except (ImportError, Exception):
            pass

        return False

    def _check_pyproject_text(
        self, pyproject_path: Path, dependencies: list[str]
    ) -> bool:
        """Check pyproject.toml using text parsing as fallback."""
        try:
            with open(pyproject_path) as f:
                content = f.read().lower()
                for dep in dependencies:
                    if dep.lower() in content:
                        return True
        except OSError:
            pass

        return False

    def _extract_dependency_name(self, dep_line: str) -> str:
        """Extract clean dependency name from dependency specification."""
        return dep_line.split(">=")[0].split("==")[0].split("[")[0].strip()

    def _check_requirements_dependencies(self, dependencies: list[str]) -> bool:
        """Check dependencies in requirements.txt file."""
        req_path = self.project_path / "requirements.txt"
        if not req_path.exists():
            return False

        try:
            with open(req_path) as f:
                content = f.read().lower()
                for dep in dependencies:
                    if dep.lower() in content:
                        return True
        except OSError:
            pass

        return False

    def get_project_info(self) -> dict[str, Any]:
        """Get comprehensive project information."""
        return {
            "primary_type": self.detect_project_type(),
            "all_types": self.detect_all_project_types(),
            "has_package_json": self._file_exists("package.json"),
            "has_pyproject_toml": self._file_exists("pyproject.toml"),
            "has_cargo_toml": self._file_exists("Cargo.toml"),
            "has_go_mod": self._file_exists("go.mod"),
            "has_git": self._file_exists(".git"),
            "has_docker": self._file_exists("Dockerfile")
            or self._file_exists("docker-compose.yml"),
            "has_tests": self._has_test_files(),
            "has_docs": self._has_documentation_files(),
        }

    def _has_test_files(self) -> bool:
        """Check if the project has test files."""
        test_patterns = [
            "test",
            "tests",
            "spec",
            "__tests__",
            "*.test.*",
            "*_test.*",
            "test_*",
            "*_spec.*",
        ]

        for pattern in test_patterns:
            if pattern.startswith("*"):
                # File pattern
                if self._has_files_with_extension(pattern):
                    return True
            else:
                # Directory pattern
                if (self.project_path / pattern).exists():
                    return True

        return False

    def _has_documentation_files(self) -> bool:
        """Check if the project has documentation."""
        doc_files = ["README.md", "README.rst", "README.txt", "docs", "documentation"]

        for doc_file in doc_files:
            if self._file_exists(doc_file) or (self.project_path / doc_file).is_dir():
                return True

        return False

File: src/project_types/detector.ts
/**
 * Enhanced project type detection with comprehensive framework support
 * TypeScript translation of vibe/project_types/detector.py
 */

import * as fs from 'fs';
import * as path from 'path';

export class ProjectDetector {
  private projectPath: string;

  constructor(projectPath = '.') {
    this.projectPath = path.resolve(projectPath);
  }

  detectProjectType(): string {
    const detections = this.detectAllProjectTypes();

    // Return the most specific/confident detection
    // Priority order: specific frameworks > language > generic
    const priorityOrder = [
      'react',
      'next',
      'nuxt',
      'svelte',
      'angular',
      'rails',
      'typescript',
      'javascript',
      'rust',
      'go',
      'web',
      'node',
      'generic',
    ];

    for (const projectType of priorityOrder) {
      if (detections.includes(projectType)) {
        return projectType;
      }
    }

    return 'generic';
  }

  detectAllProjectTypes(): string[] {
    const types: string[] = [];

    // Detect JavaScript/TypeScript frameworks and add their tech stack
    types.push(...this.detectFrontendFrameworks());

    // Detect other language projects
    types.push(...this.detectOtherLanguages());

    // Detect general project characteristics
    types.push(...this.detectGeneralTypes());

    // Remove duplicates while preserving order
    return [...new Set(types)];
  }

  private detectFrontendFrameworks(): string[] {
    const types: string[] = [];

    if (this.isReactProject()) {
      types.push('react', 'javascript', 'web', 'node');
    }

    if (this.isNextProject()) {
      types.push('next', 'react', 'javascript', 'web', 'node');
    }

    if (this.isNuxtProject()) {
      types.push('nuxt', 'javascript', 'web', 'node');
    }

    if (this.isSvelteProject()) {
      types.push('svelte', 'javascript', 'web', 'node');
    }

    if (this.isAngularProject()) {
      types.push('angular', 'typescript', 'web', 'node');
    }

    if (this.isTypescriptProject()) {
      types.push('typescript', 'javascript', 'node');
    }

    if (this.isJavascriptProject()) {
      types.push('javascript', 'node');
    }

    return types;
  }

  private detectOtherLanguages(): string[] {
    const types: string[] = [];

    if (this.isRustProject()) {
      types.push('rust');
    }

    if (this.isGoProject()) {
      types.push('go');
    }

    return types;
  }

  private detectGeneralTypes(): string[] {
    const types: string[] = [];

    if (this.isWebProject()) {
      types.push('web');
    }

    if (this.isNodeProject()) {
      types.push('node');
    }

    return types;
  }

  // Framework detection methods
  private isReactProject(): boolean {
    return (
      this.hasDependency('react') ||
      this.hasDevDependency('create-react-app') ||
      this.hasFile('react.config.js')
    );
  }

  private isNextProject(): boolean {
    return (
      this.hasDependency('next') ||
      this.hasFile('next.config.js') ||
      this.hasFile('next.config.mjs')
    );
  }

  private isNuxtProject(): boolean {
    return (
      this.hasDependency('nuxt') ||
      this.hasFile('nuxt.config.js') ||
      this.hasFile('nuxt.config.ts')
    );
  }

  private isSvelteProject(): boolean {
    return (
      this.hasDependency('svelte') ||
      this.hasFile('svelte.config.js') ||
      this.hasDevDependency('@sveltejs/kit')
    );
  }

  private isAngularProject(): boolean {
    return (
      this.hasDependency('@angular/core') ||
      this.hasFile('angular.json') ||
      this.hasDevDependency('@angular/cli')
    );
  }

  private isTypescriptProject(): boolean {
    return (
      this.hasFile('tsconfig.json') ||
      this.hasDevDependency('typescript') ||
      this.hasFiles('*.ts')
    );
  }

  private isJavascriptProject(): boolean {
    return (
      this.hasFile('package.json') ||
      this.hasFiles('*.js') ||
      this.hasFile('yarn.lock') ||
      this.hasFile('package-lock.json')
    );
  }

  private isRustProject(): boolean {
    return this.hasFile('Cargo.toml');
  }

  private isGoProject(): boolean {
    return this.hasFile('go.mod') || this.hasFiles('*.go');
  }

  private isWebProject(): boolean {
    return (
      this.hasFiles('*.html') ||
      this.hasFiles('*.css') ||
      this.hasFile('index.html') ||
      this.hasDirectory('public') ||
      this.hasDirectory('static')
    );
  }

  private isNodeProject(): boolean {
    return (
      this.hasFile('package.json') ||
      this.hasFile('node_modules') ||
      this.hasFile('yarn.lock') ||
      this.hasFile('pnpm-lock.yaml')
    );
  }

  // Utility methods
  private hasFile(filename: string): boolean {
    return fs.existsSync(path.join(this.projectPath, filename));
  }

  private hasDirectory(dirname: string): boolean {
    const dirPath = path.join(this.projectPath, dirname);
    return fs.existsSync(dirPath) && fs.statSync(dirPath).isDirectory();
  }

  private hasFiles(pattern: string): boolean {
    try {
      const files = fs.readdirSync(this.projectPath);
      const regex = new RegExp(pattern.replace('*', '.*'));
      return files.some(file => regex.test(file));
    } catch {
      return false;
    }
  }

  private hasDependency(depName: string): boolean {
    return this.hasPackageJsonDependency(depName, 'dependencies');
  }

  private hasDevDependency(depName: string): boolean {
    return this.hasPackageJsonDependency(depName, 'devDependencies');
  }

  private hasPackageJsonDependency(depName: string, section: string): boolean {
    try {
      const packageJsonPath = path.join(this.projectPath, 'package.json');
      if (!fs.existsSync(packageJsonPath)) return false;

      const packageJson = JSON.parse(fs.readFileSync(packageJsonPath, 'utf-8'));
      return packageJson[section] && depName in packageJson[section];
    } catch {
      return false;
    }
  }
}
